card/lib/card/cache.rb

```diff
@@ -96,127 +99,70 @@ def obj_to_key obj
 
       def prepopulate
         return unless @@prepopulating
+        soft = Card.cache.soft
         @@rule_cache ||= Card.rule_cache
         @@user_ids_cache ||= Card.user_ids_cache
         @@read_rule_cache ||= Card.read_rule_cache
         @@rule_keys_cache ||= Card.rule_keys_cache
-        Card.cache.write_local 'RULES', @@rule_cache
-        Card.cache.write_local 'READRULES', @@read_rule_cache
-        Card.cache.write_local 'USER_IDS', @@user_ids_cache
-        Card.cache.write_local 'RULE_KEYS', @@rule_keys_cache
+        soft.write 'RULES', @@rule_cache
+        soft.write 'READRULES', @@read_rule_cache
+        soft.write 'USER_IDS', @@user_ids_cache
+        soft.write 'RULE_KEYS', @@rule_keys_cache
       end
     end
 
-    attr_reader :prefix, :store, :klass
-    attr_accessor :local
+    attr_reader :hard, :soft
 
     def initialize opts={}
-      #warn "new cache #{opts.inspect}"
       @klass = opts[:class]
-      @store = opts[:store]
-      @local = Hash.new
-      self.system_prefix = opts[:prefix] || self.class.system_prefix(opts[:class])
-      cache_by_class[klass] = self
-    end
+      cache_by_class[@klass] = self
 
+      # hard cache mirrors the db
+      # only difference to db: it caches virtual cards
+      @hard = Persistent.new opts if opts[:store]
 
-    def system_prefix= system_prefix
-      @system_prefix = system_prefix
-      if @store.nil?
-        @prefix = system_prefix + self.class.generate_cache_id + "/"
-      else
-        @system_prefix += '/' unless @system_prefix[-1] == '/'
-        @cache_id = @store.fetch(@system_prefix + "cache_id") do
-          self.class.generate_cache_id
-        end
-        @prefix = @system_prefix + @cache_id + "/"
-      end
+      # soft cache is temporary
+      # lasts only one request/script execution/console session
+      @soft = Temporary.new
     end
 
     def read key
-      if @local.has_key? key
-        read_local key
-      elsif @store
-        write_local key, @store.read(@prefix + key)
-      end
-    end
-
-    def read_local key
-      @local[key]
-    end
-
-    def write_variable key, variable, value
-      key = @prefix + key
-      if @store && (object = @store.read key)
-        object.instance_variable_set "@#{ variable }", value
-        @store.write key, object
-      end
-      value
```

Comment:
I keep thinking we need something like `write_variable`, but I'm guessing we're not using it yet?